### DISCUSSION TITLE
Can an app determine the current color scheme in its tab/pane?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#18227] [Q&A]
Can an app determine the current color scheme in its tab/pane?

Thanks!

**Comment by lhecker** [comment]
Do you mean the name of the scheme or just the colors?

**Reply by vefatica** [reply]
The name would suffice;  so would an index (if I knew how they were indexed); so would both of those.

The settings dialog suggests that they're indexed in alphabetical order.  (?)

**Reply by lhecker** [reply]
I cannot think of any approach that we'd consider "supported" right now. What I may try to do is to query the palette via VT sequences (OSC 4), open the settings.json file as well as the defaults.json file (part of the package), and try to match them. But that doesn't cover support for unpackaged Windows Terminal. You could theoretically dig deeper of course and e.g. find the Windows Terminal window and get its corresponding executable path. None of this sounds particularly robust to me unfortunately...

**Comment by vefatica** [comment]
I have the window handle.  I (fancifully) imagined something like

`SendMessage(hWndCascacia, WM_GET_SCHEME_INDEX, 0, 0)`

How does the packaged/store version differ, with respect to color schemes, from my unzipped version?

**Reply by lhecker** [reply]
> I have the window handle. I (fancifully) imagined something like [...]

That's theoretically something we could support, but I don't think I'd want to, because it would indirectly make it an official API that cannot be easily removed anymore. I'd only want to do that if there's a lot of people asking for it.

> How does the packaged/store version differ, with respect to color schemes, from my unzipped version?

Oh, I just meant that you can't hardcode the path anymore in that case.

**Comment by vefatica** [comment]
Is there an OSC4 sequence for querying the palette?

**Reply by lhecker** [reply]
> Is there an OSC4 sequence for querying the palette?

For the palette colors, yes. The one you mentioned before: #11150
AFAIK there's no sequence to query the scheme name.

**Reply by vefatica** [reply]
I know how to set them (all 19 of them).  Is there something in [#11150](https://github.com/microsoft/terminal/issues/11150) about querying them (all 19)?  If it's there, I missed it.

**Reply by DHowett** [reply]
Yes, OSC 4.

![image](https://github.com/user-attachments/assets/e56a1502-c015-48d8-8434-832668cdca4b)

**Reply by lhecker** [reply]
Oh, I thought that #11150 was about querying them (I only glanced at it when searching for OSC 4 related issues). But yeah, adding the `?` makes it possibly to query the palette. The way I do it personally, is by just generating all 16 queries for the 16 indices and emit them to the terminal all at once, followed by a DA1. Then I read & parse stdin until I encounter the DA1 response.

**Reply by DHowett** [reply]
For both of you: you can stack all of the queries in a single request.

```
ESC ] 4 ; 1 ; ? ; 2 ; ? ; 3 ; ? ; 4 ; ? ; ... ST
```

**Reply by vefatica** [reply]
OK (thanks Dustin).  I can get that to work in CMD but it looks a little buggy (0c0c instead of 0c).

![image](https://github.com/user-attachments/assets/a3eed50d-6776-400f-bc8c-32cec4d7a229)

No joy in TCC.  I think that's because of how it gets it's input.  I've asked about workarounds.

**Reply by lhecker** [reply]
That format (0c0c) is correct. That VT sequence is specified to return either 4-, 8-, 12- (!), or 16-bit colors (I believe because of X11's weird APIs). I think 4 and 12 bits are uncommon though. You could technically properly translate between 16 and 8 bits with rounding, but you could also just chop off the lower 8 bits. Personally, I'm not entirely sure yet what approach is better.

**Reply by vefatica** [reply]
But 0c0c isn't the same number as 0c, however you look at it.  000c in place of 0c seems like what it ought to be.

**Reply by lhecker** [reply]
Unfortunately, 0xffff is meant to be equal to 0xff in that specification and `0x0c / 255.0 * 65535.0` roughly equals 0x0c0c. I kind of wish it was different, because if it worked like what you said it would be similar to the scRGB color range and allow us to specify actual HDR colors with VT sequences. Alas...

**Reply by vefatica** [reply]
OK, I understand. I just had to look at it the right way.  Can 4 hex digits (16-bit) be used when setting the palette.

**Reply by vefatica** [reply]
Dustin, I can also query the OSC  10, 11, 12 entries separately.  When I try to chain them (in an obvious way (?)) I get odd results (correct, but not all of them).  Is there a better way?

![image](https://github.com/user-attachments/assets/eea42e0f-aaa2-4759-8729-48b7be6cf95b)

**Reply by DHowett** [reply]
10-12 are strange, they don’t act like 4.

4 takes *pairs* of index and value (or query)

10-12 only take values (or queries), and they increment for every ;

so you would write

```
OSC 10 ; ? ; ? ; ? ST
```

but note that on Terminal 1.22+ you will also want to query for index 17. You can add a bunch more ;? Or you can ask for 17 specifically. :)

**Reply by j4james** [reply]
> Unfortunately, 0xffff is meant to be equal to 0xff in that specification and `0x0c / 255.0 * 65535.0` roughly equals 0x0c0c. I kind of wish it was different, because if it worked like what you said it would be similar to the scRGB color range and allow us to specify actual HDR colors with VT sequences.

@lhecker I'm not sure if I've misunderstood what you're saying, but this format is actually intended to support HDR colors. `FFFF` is only interpreted as `FF` if your terminal is limited to 8-bit color components. For terminals that support 16-bit components, `FFFF` will in fact be interpreted as `FFFF`. In both cases they're essentially max brightness, but a 16-bit terminal allows you to distinguish between 256 levels in the range `FF00` to `FFFF`, while an 8-bit terminal will consider all of those as the same intensity.

Last I checked, I think both VTE and Rxvt appeared to support 16-bit components. Or at least they were storing them that way internally, because I could round-trip 16-bit values. I'm not sure if I'd be able to tell whether they were actually rendered with that level of precision.

**Reply by lhecker** [reply]
But `FFFF` is meant to be the same brightness as e.g. `\e[48;2;255;255;255m` right? I'm not sure if there's a specification for it, but I think it's commonly assumed that the SGR colors are in sRGB, which means that OSC 4 colors would also be in sRGB? That would in turn mean that the 16-bit colors "only" add "color depth" (= more fidelity), but do not allow for HDR, which would require exceeding the dynamic range of sRGB.

**Reply by j4james** [reply]
Oh OK. I see what you're getting at now. Clearly I'm not a color expert. I was just thinking of HDR as more color depth. If you need to exceed the range of sRGB I guess we'd need a mode that could change the color space. I wonder if there any terminals that already have something like that.

**Reply by j4james** [reply]
Actually I just remembered there are a bunch of other color encoding formats supported by the `OSC 4` and `OSC 10` sequences that we could implement, like `CIELab` and `TekHVC`, and those take floating point components. Would something like that be good enough to get you the HDR range you want?

**Comment by vefatica** [comment]
Thanks, Dustin.  I should have thought of that.  I know all too well how to chain them when setting colors (i.e., don't use the 11 and the 12).

What's OSC 17?

Edit: Never mind that, it was easy to find.  Are there others besides 10, 11, 12, and 17?

My goal was to identify the scheme.  Tomorrow I'll look for a _minimal_ set of OSC 4 and OSC 10+ values that ID the scheme.  I reckon it'll be only a few.  In a C program, where do I look for the responses ... stdin?

**Reply by lhecker** [reply]
> In a C program, where do I look for the responses ... stdin?

Yep, after setting `ENABLE_VIRTUAL_TERMINAL_INPUT`.

**Reply by DHowett** [reply]
> Are there others besides 10, 11, 12, and 17?

They're all in [Thomas Dickey's excellent XTerm Control Sequences](https://www.invisible-island.net/xterm/ctlseqs/ctlseqs.html) document.

To find these OSCs, search for `Change VT100 text foreground` (it will drop you in the middle of the list).

**Reply by vefatica** [reply]
Leonard, can I get the response with ReadConsole(GetStdHandle(STD_INPUT_HANDLE) ...?

Dustin, thanks for the link.  I needed something like that.  What's the status of OSC 19?

**Comment by vefatica** [comment]
Thank you, Leonard and Dustin.  I have something that works.  I'm querying the 20 colors separately.  I can also do them all at once but so far separately is faster. I noticed something peculiar and I'd be pleased if you'd speculate on what's going on.

I do WriteConsole/WaitFor/ReadConsole 20 times (palette, FG, BG, cursor, selectBG, in that order).  If I time them individually, I always see a pattern like this (below) which is striking.  Any idea what's going on?

![image](https://github.com/user-attachments/assets/c28965fe-927b-4b7e-9042-be04624e9d06)

**Reply by DHowett** [reply]
If you shuffle the order, is it still always the first five?

Can I also see the measurements for when you put all the OSC 4 requests in a single packet? I'd love to empirically observe that it is slower...

**Reply by vefatica** [reply]
OK, Here I've changed the order (20 sinfle queries) from palette-defaults_and_cursor-selectBG to defaults_and_cursor-palette-selectBG.  It'll take a while to implement doing the palette with one request.

![image](https://github.com/user-attachments/assets/8052b155-a4c7-473b-b54f-a7de89c839d7)

**Reply by vefatica** [reply]
That didn't take long!  Here's the result when I query the whole palette with one request, then the other 4.

![image](https://github.com/user-attachments/assets/b7e51775-d790-4e4c-963b-e9571465c4bc)

FWIW, I'm querying the palette like this and the other 4 similarly.  hIn and hOut are GetStdHandle(...).

```
	Timer(1);
	WriteConsole(hOut, szQueryAll, nQueryAllLength, &dwWritten, 0);
	WaitForSingleObject(hIn, 100);
	ReadConsole(hIn, szResponseAll, 600, &dwRead, 0);
	Timer(0);
```

**Reply by DHowett** [reply]
That seems strictly shorter than the sum of the 16 individual measurements...

The first 16 reads take a total of 1503 microseconds, whereas the single 16-item read takes 429 microseconds.

You're paying a couple cross-process hops and a kernel syscall transition for *every write* and *every read*, so it behooves you to batch them up.

You could also put the entire OSC for all the other palette reads into the same packet.

If you write one huge buffer that contains...

```
\e]4;0;?;1;?;2;?;3;?;4;?;5;?;6;?;7;?;8;?;9;?;10;?;11;?;12;?;13;?;14;?;15;?\e\\\e]10;?;?;?;;;;;?\e\\
```

You'll get them all at once and back to back with only two syscalls.

**Reply by vefatica** [reply]
For all of them I was using something close to that.

```
	INT nQueryAllLength = wsprintf(szQueryAll,
		OSC L"4;0;?;1;?;2;?;3;?;4;?;5;?;6;?;7;?;8;?;9;?;10;?;11;?;12;?;13;?;14;?;15;?" ST1 OSC L"10;?;?;?" ST1 OSC L"17;?" ST1);
```

The slowness of the all-at-once approach, which I reported earlier, may have been due to my parsing the response ... swscanf_s in the one-at-a-time approach and plowing through the response with wcspbrk/wcstoul in the all-at-once approach.

If I'm doing them all at once is there a chance that I'll get only a partial result when I read hIn?

**Reply by DHowett** [reply]
Only if you don't allocate enough buffer!

**Reply by vefatica** [reply]
So the responses sent with one write operation ... eh?

**Reply by DHowett** [reply]
Yep! They sure are.

**Reply by vefatica** [reply]
Thanks.  I'll go back to the all-in-one approach and try to come up with a faster way to parse the response.

It's still a mystery why the responses get quicker.

**Reply by vefatica** [reply]
That ain't so bad after all. Here's timing just the write/read.

![389122270-a3d86bc0-fda4-4f4f-bba1-d11f47b73a75](https://github.com/user-attachments/assets/90fd742b-d18b-4114-ba5b-bb10aad7066a)


And here's timing the write/read/parse.

![389122535-f2b095b1-6cb8-4fb5-8198-34b546f0e99b](https://github.com/user-attachments/assets/431ff429-3229-4e6e-9d81-eaf8efcef938)

**Reply by vefatica** [reply]
Hmmm!  I said

> If I'm doing them all at once is there a chance that I'll get only a partial result when I read hIn?


and @DHowett said

> Only if you don't allocate enough buffer!

But occasionally,

![image](https://github.com/user-attachments/assets/ec4ad68d-96a5-493c-a888-6a90f460e743)

**Reply by DHowett** [reply]
Can you dump the contents of the short reads?

**Reply by vefatica** [reply]
It's the response to the first query.  FWIW, I made a standalone EXE to do the same thing.  It misbehaves similarly.

![image](https://github.com/user-attachments/assets/2d5d715a-be29-4d26-bdd6-1b71057ce220)

**Reply by vefatica** [reply]
Apparently, and only occasionally, `WaitForSingleObject(hIn, 100);` was not long enough (and I wasn't flushing hIn).  So weird things were happening.  I can't reproduce the bad behavior using a timeout of 200 and flushing hIn appropriately ... in either the _THEME TCC plugin variable or an EXE.  I just might wrap the whole query in while `( dwRead != 522)`.

FWIW, I can easily switch between using standard handles and CONIN$/CONOUT$ handles.  Both behave/misbehave the same with respect to the current discussion, but the latter deals with externally induced redirection and piping better.

**Reply by vefatica** [reply]
When I use the same query,

`OSC L"4;0;?;1;?;2;?;3;?;4;?;5;?;6;?;7;?;8;?;9;?;10;?;11;?;12;?;13;?;14;?;15;?" ST OSC L"10;?;?;?;;;;;?" ST`

in CMD, running in OpenConsole.exe, the response doesn't include `OSC 17`.  Is that expected?

**Reply by DHowett** [reply]
Depends on the version. 17 is newly supported in 1.22.

If you parse the responses properly, you can take a missing 10, 11, 12 or 17 in stride. Any or all of them could be omitted for various reasons.

**Reply by vefatica** [reply]
I'm using WindowsTerminal 1.22.2411.18002 (preview) , installed two days ago, and the OpenConsole.exe that comes with it.

**Reply by DHowett** [reply]
Oh, right, yes. It's coming back to me.

The foreground, background, cursor and selection colors will *only* be included if they are representable in an X11 XParseColor compatible format.

The idea is that an application can request a palette readout and use the response it gets to understand or completely restore the palette.

Some colors, such as "acrylic background" (Terminal) and "inverse video cursor/selection" cannot be represented. They will be ignored.

Conhost's default selection color is "inverse video." It will not be returned from `OSC 17` unless an actual color is set _with_ `OSC 17`.

**Reply by vefatica** [reply]
Thanks.  I see.  Should it actually work? [Selection background shown below]

![image](https://github.com/user-attachments/assets/f4506cb0-c60a-489f-a192-98a5f50cd208)

**Reply by DHowett** [reply]
This is one of those weird things where we only implemented it in the modern text rendering engine. Conhost doesn't use that one by _default,_ but it can be configured to.

**Reply by vefatica** [reply]
"Conhost" ... as a role for OpenConsole.exe, right? ... not conhost.exe.

**Reply by DHowett** [reply]
_every time I say it, openconsole is conhost._ it is the console host. it operates as the console host. it contains the code of the console host (and the console host contains the code of openconsole.)

**Reply by DHowett** [reply]
(Given that I can't go retroactively update the version called `conhost.exe` in the OS without getting mummified in a bunch of red tape, I'll keep calling all of the versions "conhost" :))

**Reply by vefatica** [reply]
I'd replace the OS's conhost.exe if it weren't for the white scroll bar.

**Reply by lhecker** [reply]
> I'd replace the OS's conhost.exe if it weren't for the white scroll bar.

That bears some significant risk in case that the OpenConsole version you're installing has a bug, so we always tell people to never do that.
That said, if you're search for a branch named after "dark mode" you'll find one that features the necessary, not-really-secret-but-still-forbidden code to make it work. You can compile and use it if you'd like, but we recommend against doing that. _But_ if you do it anyway, please make a backup of your original conhost.exe and let us know of any weird issues that you encounter.

**Reply by vefatica** [reply]
I'd love to do that and I don't really have a clue how to.  And I might not have all the VS tools necessary.  I tried opening OpenConsole.sln and got this.

![image](https://github.com/user-attachments/assets/de319637-eecd-46c2-b347-e1046eba25b5)

This is what I have.

![image](https://github.com/user-attachments/assets/ac291d78-9ff0-43c7-89f6-f51bcbb62774)

**Reply by lhecker** [reply]
That's a warning I haven't seen yet. Your screenshot doesn't show what modifications it wants to do, unfortunately.
If you open a PowerShell prompt in the root of the project (ideally a modern PowerShell 7 one, because that's what we primarily test for), you can do this:
```powershell
Import-Module .\tools\OpenConsole.psm1
Set-MsbuildDevEnvironment
Invoke-OpenConsoleBuild '-p:Configuration=Release' '-t:Conhost\Host_EXE'
```

It should take <2min and the result will be at `.\bin\x64\Release\OpenConsole.exe`.

**Reply by vefatica** [reply]
P:\terminal-1.22.2362.0\terminal-1.22.2362.0> Set-MsbuildDevEnvironment
Import-LocalModule: The term 'git' is not recognized as a name of a cmdlet, function, script file, or executable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

**Reply by DHowett** [reply]
Okay. It is fair to assume that you do not have a normal development environment as seen on most development workstations post-2010.

Probably safe to cut our losses early :)

**Reply by vefatica** [reply]
What's normal?  I installed what I thought I'd use.

**Reply by vefatica** [reply]
The warning also had this.

![image](https://github.com/user-attachments/assets/9a54dd2f-c192-4c7a-a9fc-e14e0503b583)

I couldn't find a log.  "Copy information" gave only this.  I don't think any changes were made.

> Non-functional changes required
> Visual Studio will automatically make non-functional changes to the following projects in order to enable them to open in Visual Studio 2015, Visual Studio 2013, Visual Studio 2012, and Visual Studio 2010 SP1. Project behavior will not be impacted.
> 	 - OpenConsole, "P:\terminal-1.22.2362.0\terminal-1.22.2362.0\OpenConsole.sln"

**Reply by vefatica** [reply]
I got a migration report.  These are "normal"?

![image](https://github.com/user-attachments/assets/0a7a94fd-d688-429e-894b-a778c2b91160)

**Reply by vefatica** [reply]
Whew!  That was a chore.  But now I have

![image](https://github.com/user-attachments/assets/3350f382-0393-4661-99ae-424b352d2f49)

I don't even know what a "branch" is.  How do I search for a "dark mode" branch and what do I do with whatever I find?
